### ISSUE TITLE
Shader Debugging is not supported on OpenGL

### ISSUE THREAD

**Issue by karyon** [baldurk/renderdoc#293]
When clicking on the "Debug Compute Shader", literally nothing happens.

I have a GTX 750 Ti, driver version 364.72, Windows 10, using OpenGL. i also tested it with another random OpenGL app (tesseract) i had on the machine.

**Comment by baldurk** [comment]
Yes, shader debugging and pixel history are both only supported on D3D11 currently. This should be made more obvious in the UI though when you're using a different API.

Would you mind opening a separate bug for one of them (either debugging or pixel history)? Since both features are a large task I'd rather track them separately.

**Comment by karyon** [comment]
yeah sure.

**Comment by chrisfairc** [comment]
This would be awesome!

**Comment by Inori** [comment]
there's a [spirv-vm](https://github.com/dfranx/SPIRV-VM) by @dfranx
can it be added to renderdoc?

**Comment by jnhyatt** [comment]
DirectX has gotten me used to SPIR-V debugging. GLSL debugging would be nice, but at this point, anything would make me very happy. Is this on the roadmap? Perhaps for 1.10...?

**Comment by baldurk** [comment]
That makes no sense, DirectX doesn't use SPIR-V.

Shader debugging for OpenGL may happen at some point now since SPIR-V debugging has been implemented for vulkan, but certainly not in the next release. It is still quite low priority due to OpenGL being a flawed API not suited to shader debugging.

The most that could be supported would be debugging GL shaders that are either already SPIR-V - which is extremely rare - or modern enough GLSL that they can be compiled to SPIR-V - which is still quite rare as the large majority of people using OpenGL are using old/legacy shaders. Those who use modern shaders likely don't use OpenGL. That is partly why it's low priority since it would be only implementing partial support for a subset of users.

**Comment by Trass3r** [comment]
> GLSL that they can be compiled to SPIR-V

Looks like RenderDoc is already doing this for reflection purposes. Could that be reused for debugging?

**Comment by listevemeta** [comment]
What's the current status for OpenGL Pixel History and Shader debugging? I am trying it on a VR app on Quest2. I noticed that the pixel history button only shows me the glClear() call in the list, but never shows me a glDraw call.

**Comment by baldurk** [comment]
This issue is about shader debugging on OpenGL, which is not supported. If you are encountering a bug with using the pixel history please file an issue about it and follow the issue template for a bug report, do not report it here.

**Comment by bryanedds** [comment]
I'd like to investigate the scope of effort required for an individual contributor such as myself to implement this feature.

Any advice on how to get started on that? I'm at the point where I can make this my first priority, nearly regardless of how long it takes. But before getting started, I'd like to know what I'm getting myself into and if any guidance would be available from the maintainers in the process :)

**Comment by baldurk** [comment]
The effort would be quite high, and I would not recommend this as your first contribution to the project as it will be quite complex to implement. The most feasible implementation would be as mentioned above to restrict it only to shaders that can be compiled to SPIR-V using glslang and expand and re-use the SPIR-V debugging from Vulkan to provide the debug information.

Getting started would be to familiarise yourself with RenderDoc's codebase and how the Vulkan shader debugging works particularly with how it interfaces with the API and what things would need to be abstacted and re-implemented on GL. Again this would not be a small project and I would suggest attempting something smaller at first if you are a new contributor. If you do start working on this please stay in contact via email (easier than very noisy comments here) as suggested in the CONTRIBUTING documentation to ensure you don't do a huge amount of work only to find that it needs to be rewritten when it comes time to review.

**Comment by bryanedds** [comment]
Hello! I appreciate the in-depth response!

Yes, the first thing I'm wanting to attempt for this particular issue is to pursue the SPIR-V / glslang approach you mention. This should at least satisfy our team's own use case since we're going to change our engine in tandem to use glslang for our opengl shaders.

I've already started taking an overview of the existing code base.

I'll follow up with an e-mail as we determine how to proceed!

**Comment by bryanedds** [comment]
From our cursory scan, it's looking like this sort of endeavor is on the order of many 10klocs - even bigger than we had anticipated! While that was going on, we just found some rendering issues on Intel GPUs with OpenGL 3D renderer. Coupled with the issues we're currently having with AMD OpenGL, we're now looking at instead increasing the priority of our Vulkan renderer rewrites. If we go ahead and do that, we'll probably end up having to drop further effort into this here. Once we firm up things further, we'll provide another update.

UPDATE: Given the scope of the task, **it's looking like we can't currently commit to this.** Strong encouragement for anyone who can make the time, tho. Say what you want about OpenGL, but its much smaller and higher-level API still offers a significant value proposition for smaller / underfunded game devs.

**Comment by Trass3r** [comment]
> restrict it only to shaders that can be compiled to SPIR-V using glslang and expand and re-use the SPIR-V debugging from Vulkan

> it's looking like this sort of endeavor is on the order of many 10klocs

Could you give a summary of your insights? 
Is the effort also this big if you restrict it to [Vulkan GLSL](https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_vulkan_glsl.txt)?

**Comment by bryanedds** [comment]
> Could you give a summary of your insights?

I'm afraid most of the temporary cursory understanding I had built for this analysis has now been mostly faded from reliable memory.

> Is the effort also this big if you restrict it to [Vulkan GLSL](https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_vulkan_glsl.txt)?

IIRC, yes.

**Comment by baldurk** [comment]
A limited form of shader debugging is now supported on OpenGL. Although this does not support all shaders on all drivers I am still closing this issue as expanding the support any further is likely so low in terms of value-vs-cost that I would consider it a rejected feature.

The most significant limitation is that the shader _can't_ be so outdated and legacy that it is incompatible with SPIR-V compilation. Allowance is made via glslang's auto-mapping functionality for texture/buffer uniforms that don't have explicit bindings, and inputs/outputs/value uniforms without explicit locations. However in some of these cases it is possible the debugging may fail so it is expected that the most reliable results will come from shaders that compile as-is to SPIR-V without modification.

Note that as a result of the compilation this may expose undefined behaviour in the shader depending on how different the generated shader is from the driver compiled GLSL and differences in reflected bindings. The effects of this may vary and best results will come from shaders that are explicitly laid out and explicitly fully bound compared to shaders that are lazily bound.

Vertex, pixel and compute shaders can be debugged, tessellation and geometry stages are not supported as with all other APIs.

The feature also has high minimum requirements - both `GL_ARB_compute_shader` and `ARB_shader_storage_buffer_object` (or equivalent core version functionality) - so it's expected that not all GL drivers will be supported. Shader debugging will be disabled entirely for any OpenGL capture if the minimum requirements are not met. It also requires SSBO support for the stage to be debugged, meaning that commonly on Android vertex shaders will not be debuggable.

This information is also in the documentation under the details of the OpenGL & OpenGL ES Support.
